### ISSUE TITLE
Crash: StringIndexOutOfBoundsException in LatexAnnotatedTextBuilder

### ISSUE THREAD

**Issue by theFSO** [valentjn/vscode-ltex#236]
Note: It is highly recommended to follow the instructions at https://valentjn.github.io/vscode-ltex/docs/contributing-code-issues.html#how-to-report-bugs and use the `LTeX: Report bug in LTeX` command from within Visual Studio Code. Per the contribution guidelines, deleting parts of the template or not filling in vital information may result in the issue to be immediately closed as invalid.

**Describe the bug**
LTex is not working properly, when executing Ltex:check current doc, it out put internal error

**Steps to reproduce**

1. No idea this is the first time install Ltex for me and it just not working.

**Expected behavior**
How to fix it

**Sample document**
If the bug occurs for a specific document (e.g. LaTeX), please paste it here. If your document is very long or confidential, please create and attach a smaller example for which the bug still occurs so that we can reproduce it.

<details>

```
REPLACE_THIS_WITH_SAMPLE_DOCUMENT
```

</details>

**LTeX configuration**
Please paste all configuration settings starting with `ltex.` from your `settings.json`. You can help us by temporarily removing some irrelevant settings from your `settings.json` and see if the bug still occurs.

<details>

```
    "ltex.enabled": ["bibtex", "latex", "markdown", "rsweave"],
    "ltex.statusBarItem": true,
    "ltex.trace.server": "verbose",
    "ltex.additionalRules.enablePickyRules": true
```

</details>

**"LTeX Language Server" log file**
First, reproduce the bug. Then, go to `View` → `Output` and select `LTeX Language Server` in the drop-down list. Paste this log here:

<details>

```
Jan 28, 2021 3:22:17 AM org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint notify
INFO: Unsupported notification method: $/setTraceNotification
[Error - 3:22:23 AM] Registering progress handler for token {"uri":"file:///home/ubuntu/cs2me3/guoz59/A1/report/report.tex","operation":"checkDocument","counter":0} failed.
Error: Progress handler for token {"uri":"file:///home/ubuntu/cs2me3/guoz59/A1/report/report.tex","operation":"checkDocument","counter":0} already registered
	at Object.onProgress (/home/ubuntu/.vscode-server/extensions/valentjn.vscode-ltex-8.3.0/dist/extension.js:1:188606)
	at P.onProgress (/home/ubuntu/.vscode-server/extensions/valentjn.vscode-ltex-8.3.0/dist/extension.js:1:243094)
	at new t.ProgressPart (/home/ubuntu/.vscode-server/extensions/valentjn.vscode-ltex-8.3.0/dist/extension.js:1:282879)
	at /home/ubuntu/.vscode-server/extensions/valentjn.vscode-ltex-8.3.0/dist/extension.js:1:282605
	at /home/ubuntu/.vscode-server/extensions/valentjn.vscode-ltex-8.3.0/dist/extension.js:1:183064
	at /home/ubuntu/.vscode-server/extensions/valentjn.vscode-ltex-8.3.0/dist/extension.js:1:184106
	at Immediate._onImmediate (/home/ubuntu/.vscode-server/extensions/valentjn.vscode-ltex-8.3.0/dist/extension.js:1:186491)
	at processImmediate (internal/timers.js:439:21)
Jan 28, 2021 3:22:23 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 0
java.util.concurrent.CompletionException: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 0
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:645)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:212)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 0
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3319)
	at java.base/java.lang.String.substring(String.java:1874)
	at org.bsplines.ltexls.parsing.latex.LatexAnnotatedTextBuilder.addCode(LatexAnnotatedTextBuilder.java:510)
	at org.bsplines.ltexls.parsing.latex.LatexAnnotatedTextBuilder.addCode(LatexAnnotatedTextBuilder.java:26)
	at org.bsplines.ltexls.server.DocumentChecker.buildAnnotatedTextFragments(DocumentChecker.java:54)
	at org.bsplines.ltexls.server.DocumentChecker.check(DocumentChecker.java:193)
	at org.bsplines.ltexls.server.LtexTextDocumentItem.lambda$check$6(LtexTextDocumentItem.java:471)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	... 12 more
```

</details>

**"LTeX Language Client" log file**
First, set the `ltex.trace.server` setting in your `settings.json` to `"verbose"`. Then, reload the VS Code window and reproduce the bug. Go to `View` → `Output` and select `LTeX Language Client` in the drop-down list. Paste this log here (note: it will contain your checked document):

<details>

```
[Trace - 3:22:23 AM] Received response 'workspace/executeCommand - (28)' in 68ms. Request failed: Internal error. (-32603).
Error data: "java.util.concurrent.CompletionException: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 0\n\tat java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)\n\tat java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:645)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)\n\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:212)\n\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)\n\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)\n\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)\n\tat org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\nCaused by: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 0\n\tat java.base/java.lang.String.checkBoundsBeginEnd(String.java:3319)\n\tat java.base/java.lang.String.substring(String.java:1874)\n\tat org.bsplines.ltexls.parsing.latex.LatexAnnotatedTextBuilder.addCode(LatexAnnotatedTextBuilder.java:510)\n\tat org.bsplines.ltexls.parsing.latex.LatexAnnotatedTextBuilder.addCode(LatexAnnotatedTextBuilder.java:26)\n\tat org.bsplines.ltexls.server.DocumentChecker.buildAnnotatedTextFragments(DocumentChecker.java:54)\n\tat org.bsplines.ltexls.server.DocumentChecker.check(DocumentChecker.java:193)\n\tat org.bsplines.ltexls.server.LtexTextDocumentItem.lambda$check$6(LtexTextDocumentItem.java:471)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)\n\t... 12 more\n"

```

</details>

**Version information**
List here the version information of the relevant software.

- Operating system: Linux (which distribution/version), macOS xx.xx, or Windows xx
- VS Code: 1.52.1
- vscode-ltex: 8.3.0
- ltex-ls: x.xx (only if not using ltex-ls automatically downloaded by LTeX)
- Java: x.xx (usually obtained with `java -version`, only if not using Java automatically downloaded by LTeX)

**Additional context/information**
You can add any other context or information about the problem here.

**Comment by valentjn** [comment]
Thanks for the report. You have a `\begin` or `\end` without an argument (`{environmentname}`) somewhere in your document. As a workaround, add the argument and reload the VS Code window (`Ctrl+P` → `Developer: Reload Window`).

**Comment by valentjn** [comment]
Fix released in 8.4.0.
